File: web/src/lib/ai/matching.ts
import { cosineSimilarity, embed } from "./embeddings";

const HARD_WEIGHT = 0.6;
const SEMANTIC_WEIGHT = 0.4;

export type MatchResult = {
  match_percentage: number;
  hard_skill_overlap: number;
  semantic_fit: number;
};

export function hardSkillOverlap(studentSkills: string[], requiredSkills: string[]): number {
  const required = new Set(requiredSkills.map((s) => s.trim().toLowerCase()).filter(Boolean));
  if (required.size === 0) return 0;
  const student = new Set(studentSkills.map((s) => s.trim().toLowerCase()).filter(Boolean));
  let overlap = 0;
  for (const skill of required) if (student.has(skill)) overlap++;
  return Math.round((overlap / required.size) * 10000) / 100;
}

export async function semanticFit(studentBio: string, opportunityDescription: string): Promise<number> {
  const [studentVec, opportunityVec] = await Promise.all([embed(studentBio || ""), embed(opportunityDescription || "")]);
  const sim = cosineSimilarity(studentVec, opportunityVec);
  return Math.round(Math.max(0, sim) * 10000) / 100;
}

export async function computeMatch(
  studentSkills: string[],
  studentBio: string,
  opportunityRequirements: string[],
  opportunityDescription: string
): Promise<MatchResult> {
  const hard = hardSkillOverlap(studentSkills, opportunityRequirements);
  const semantic = await semanticFit(studentBio, opportunityDescription);
  const total = Math.round((HARD_WEIGHT * hard + SEMANTIC_WEIGHT * semantic) * 100) / 100;
  return { match_percentage: total, hard_skill_overlap: hard, semantic_fit: semantic };
}

File: engine/app/services/matching.py
from app.services.embeddings import cosine_similarity, embed

HARD_WEIGHT = 0.60
SEMANTIC_WEIGHT = 0.40


def hard_skill_overlap(student_skills: list[str], required_skills: list[str]) -> float:
    """S_hard = |K_student ∩ K_required| / |K_required|, as a 0-100 percentage."""
    required = {s.strip().lower() for s in required_skills if s.strip()}
    if not required:
        return 0.0
    student = {s.strip().lower() for s in student_skills if s.strip()}
    overlap = student & required
    return round(100 * len(overlap) / len(required), 2)


async def semantic_fit(student_bio: str, opportunity_description: str) -> float:
    """S_semantic = cosine(profile embedding, opportunity embedding), as 0-100."""
    student_vec = await embed(student_bio or "")
    opportunity_vec = await embed(opportunity_description or "")
    sim = cosine_similarity(student_vec, opportunity_vec)
    return round(max(0.0, sim) * 100, 2)


async def compute_match(
    student_skills: list[str],
    student_bio: str,
    opportunity_requirements: list[str],
    opportunity_description: str,
) -> dict:
    hard = hard_skill_overlap(student_skills, opportunity_requirements)
    semantic = await semantic_fit(student_bio, opportunity_description)
    total = round(HARD_WEIGHT * hard + SEMANTIC_WEIGHT * semantic, 2)
    return {
        "match_percentage": total,
        "hard_skill_overlap": hard,
        "semantic_fit": semantic,
    }
